File: WCF/Windows SDK WCF Samples/TechnologySamples Beta 1/Scenario/LoadBalancer/VB/library/LoadBalancerBehaviorSection.vb
' Copyright (c) Microsoft Corporation.  All Rights Reserved.

Imports System
Imports System.ServiceModel.Channels
Imports System.Configuration
Imports System.Globalization
Imports System.ServiceModel
Imports System.ServiceModel.Configuration

Namespace Microsoft.ServiceModel.Samples

    Public Class LoadBalancerBehaviorElement
        Inherits BehaviorExtensionElement

        Const AddressString As String = "address"
        Const BindingConfigurationString As String = "bindingConfiguration"
        Const BindingString As String = "binding"
        Const BindingsSectionGroupPath As String = "system.serviceModel/bindings"

        <ConfigurationProperty(LoadBalancerBehaviorElement.AddressString, DefaultValue:="")> _
        Public Property Address() As Uri

            Get

                Return DirectCast(MyBase.Item(LoadBalancerBehaviorElement.AddressString), Uri)

            End Get
            Set(ByVal value As Uri)

                MyBase.Item(LoadBalancerBehaviorElement.AddressString) = value

            End Set

        End Property

        <ConfigurationProperty(LoadBalancerBehaviorElement.BindingConfigurationString, DefaultValue:="")> _
        <StringValidator(MinLength:=0)> _
        Public Property BindingConfiguration() As String

            Get

                Return DirectCast(MyBase.Item(LoadBalancerBehaviorElement.BindingConfigurationString), String)

            End Get
            Set(ByVal value As String)

                If String.IsNullOrEmpty(value) Then

                    value = String.Empty

                End If
                MyBase.Item(LoadBalancerBehaviorElement.BindingConfigurationString) = value

            End Set

        End Property

        <ConfigurationProperty(LoadBalancerBehaviorElement.BindingString, Options:=ConfigurationPropertyOptions.IsRequired)> _
        <StringValidator(MinLength:=0)> _
        Public Property Binding() As String

            Get

                Return DirectCast(MyBase.Item(LoadBalancerBehaviorElement.BindingString), String)

            End Get
            Set(ByVal value As String)

                If String.IsNullOrEmpty(value) Then

                    value = String.Empty

                End If
                MyBase.Item(LoadBalancerBehaviorElement.BindingString) = value

            End Set

        End Property

        Private Shared Function GetSection(ByVal sectionPath As String) As Object

            Dim retval As Object = ConfigurationManager.GetSection(sectionPath)
            If retval Is Nothing Then

                Throw New ConfigurationErrorsException("Configuration section not found")

            End If
            Return retval

        End Function

        Private Shared Function LookupBinding(ByVal binding As String, ByVal configurationName As String) As Binding

            If String.IsNullOrEmpty(binding) Then

                Throw New ConfigurationErrorsException("Binding type cannot be null or empty.")

            End If

            Dim bindingsSection As BindingsSection = DirectCast(ConfigurationManager.GetSection(LoadBalancerBehaviorElement.BindingsSectionGroupPath), BindingsSection)
            Dim bindingCollectionElement As BindingCollectionElement = bindingsSection(binding)

            Dim retval As Binding = DirectCast(System.Activator.CreateInstance(bindingCollectionElement.BindingType), Binding)
            If Not String.IsNullOrEmpty(configurationName) Then

                ' We are looking for a specific instance, not the default. 
                Dim configuredBindingFound As Boolean = False

                ' The Bindings property is always public
                For Each configElement As Object In bindingCollectionElement.ConfiguredBindings

                    Dim bindingElement As IBindingConfigurationElement = TryCast(configElement, IBindingConfigurationElement)
                    If bindingElement IsNot Nothing Then

                        If bindingElement.Name.Equals(configurationName, StringComparison.CurrentCulture) Then

                            bindingElement.ApplyConfiguration(retval)
                            configuredBindingFound = True

                        End If

                    End If

                Next
                If Not configuredBindingFound Then

                    ' We expected to find an instance, but didn't.
                    ' Return null. 
                    retval = Nothing

                End If

            End If
            Return retval

        End Function

        Protected Overloads Overrides Function CreateBehavior() As Object

            Dim loadBalancerAddress As New EndpointAddress(DirectCast(MyBase.Item(LoadBalancerBehaviorElement.AddressString), Uri))

            Dim bindingType As String = DirectCast(MyBase.Item(LoadBalancerBehaviorElement.BindingString), String)
            Dim configName As String = DirectCast(MyBase.Item(LoadBalancerBehaviorElement.BindingConfigurationString), String)
            Dim binding As Binding = LookupBinding(bindingType, configName)

            If binding Is Nothing Then
                Throw New ConfigurationErrorsException("Could not find specified binding.")
            End If

            Return New LoadBalancerBehavior(loadBalancerAddress, binding)

        End Function

        Public Overloads Overrides ReadOnly Property BehaviorType() As Type

            Get

                Return GetType(LoadBalancerBehavior)

            End Get

        End Property

    End Class

End Namespace

File: WCF/Windows SDK WCF Samples/TechnologySamples Beta 1/Scenario/LoadBalancer/CS/library/LoadBalancerBehaviorSection.cs

//  Copyright (c) Microsoft Corporation.  All Rights Reserved.

using System;
using System.ServiceModel.Channels;
using System.Configuration;
using System.Globalization;
using System.ServiceModel;
using System.ServiceModel.Configuration;

namespace Microsoft.ServiceModel.Samples
{
    public class LoadBalancerBehaviorElement : BehaviorExtensionElement
    {
        const string AddressString = "address";
        const string BindingConfigurationString = "bindingConfiguration";
        const string BindingString = "binding";
        const string BindingsSectionGroupPath = "system.serviceModel/bindings";

        [ConfigurationProperty(LoadBalancerBehaviorElement.AddressString, DefaultValue = "")]
        public Uri Address
        {
            get { return (Uri)base[LoadBalancerBehaviorElement.AddressString]; }
            set { base[LoadBalancerBehaviorElement.AddressString] = value; }
        }

        [ConfigurationProperty(LoadBalancerBehaviorElement.BindingConfigurationString, DefaultValue = "")]
        [StringValidator(MinLength = 0)]
        public string BindingConfiguration
        {
            get { return (string)base[LoadBalancerBehaviorElement.BindingConfigurationString]; }
            set
            {
                if (string.IsNullOrEmpty(value))
                {
                    value = string.Empty;
                }
                base[LoadBalancerBehaviorElement.BindingConfigurationString] = value;
            }
        }

        [ConfigurationProperty(LoadBalancerBehaviorElement.BindingString, Options = ConfigurationPropertyOptions.IsRequired)]
        [StringValidator(MinLength = 0)]
        public string Binding
        {
            get { return (string)base[LoadBalancerBehaviorElement.BindingString]; }
            set
            {
                if (string.IsNullOrEmpty(value))
                {
                    value = string.Empty;
                }
                base[LoadBalancerBehaviorElement.BindingString] = value;
            }
        }

        static object GetSection(string sectionPath)
        {
            object retval = ConfigurationManager.GetSection(sectionPath);
            if (retval == null)
            {
                throw new ConfigurationErrorsException("Configuration section not found");
            }
            return retval;
        }

        static Binding LookupBinding(string binding, string configurationName)
        {
            if (string.IsNullOrEmpty(binding))
            {
                throw new ConfigurationErrorsException("Binding type cannot be null or empty.");
            }

            BindingsSection bindingsSection = (BindingsSection)ConfigurationManager.GetSection(LoadBalancerBehaviorElement.BindingsSectionGroupPath);
            BindingCollectionElement bindingCollectionElement = bindingsSection[binding];


            Binding retval = (Binding)System.Activator.CreateInstance(bindingCollectionElement.BindingType);
            if (!string.IsNullOrEmpty(configurationName))
            {
                // We are looking for a specific instance, not the default. 
                bool configuredBindingFound = false;

                // The Bindings property is always public
                foreach (object configElement in bindingCollectionElement.ConfiguredBindings)
                {
                    IBindingConfigurationElement bindingElement = configElement as IBindingConfigurationElement;
                    if (bindingElement != null)
                    {
                        if (bindingElement.Name.Equals(configurationName, StringComparison.CurrentCulture))
                        {
                            bindingElement.ApplyConfiguration(retval);
                            configuredBindingFound = true;
                        }
                    }
                }
                if (!configuredBindingFound)
                {
                    // We expected to find an instance, but didn't.
                    // Return null. 
                    retval = null;
                }
            }
            return retval;
        }

        protected override object CreateBehavior()
        {
            EndpointAddress loadBalancerAddress = new EndpointAddress((Uri)base[LoadBalancerBehaviorElement.AddressString]);

            string bindingType = (string)base[LoadBalancerBehaviorElement.BindingString];
            string configName = (string)base[LoadBalancerBehaviorElement.BindingConfigurationString];
            Binding binding = LookupBinding(bindingType, configName);

            if (binding == null)
                throw new ConfigurationErrorsException("Could not find specified binding.");

            return new LoadBalancerBehavior(loadBalancerAddress, binding);
        }

        public override Type BehaviorType
        {
            get { return typeof(LoadBalancerBehavior); }
        }
    }
}
